### ISSUE TITLE
fitting discrete distributions, hmm transition matrix not changing

### ISSUE THREAD

**Issue by mabuodeh** [jmschrei/pomegranate#673]
Hey Jacob!

I created an hmm as follows:
1. I first obtained a list of observations and states
2. created a dict, where the keys are the observations, and the value is a uniform number (for all observations; 1/len(observations)) -> observation_dict = DiscreteDistribution({ "obs1": 0.25, "obs2":0.25 ... })
3. for every state, i created a State object -> State(observation_dict, name=state_name)
4. I added all the states to the hmm, then added transitions from model.start to all the states, as well as all the states to each other. the number assigned for the transitions is also uniform (1/len(states)).
5. baked the model

after these steps, I fit the model using a training set, with the set's labels:
model.fit(sequences=X_train, labels=y_train, algorithm='baum-welch', use_pseudocount=True, max_iterations=10)
Where X_train is the lists of observations, and y_train are their respective lists of labels.

When I print the model, the observation probabilities change, but the transition matrix is still the original number I assigned at the beginning.

Am I doing something wrong?

Thanks in advance!

**Comment by jmschrei** [comment]
Howdy

Would it be possible to provide some example code that I could run that shows this issue?

**Comment by mabuodeh** [comment]
`
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn; seaborn.set_style('whitegrid')
import numpy

from pomegranate import *

numpy.random.seed(0)
numpy.set_printoptions(suppress=True)

# %load_ext watermark
# %watermark -m -n -p numpy,scipy,pomegranate

def generate_uniform_distribution(observations):
    template_dist = {}
    
    obs_len = 1 / len(observations)
    for obs in observations:
        template_dist[obs] = obs_len

    return template_dist


def generate_state_with_uniform_distribution(state_name, observations):
    # d1 = DiscreteDistribution({'A': 0.25, 'C': 0.25, 'G': 0.25, 'T': 0.25})
    temp_dist = DiscreteDistribution(generate_uniform_distribution(observations))

    temp_state = State(temp_dist, name=state_name)
    return temp_state


def get_uniform_dist_for_all_states(unique_states, unique_syscalls):
    ret_states = []
    for st in unique_states:
        temp_dist = DiscreteDistribution(generate_uniform_distribution(unique_syscalls))
        temp_state = State(temp_dist, name=st)
        ret_states.append(temp_state)

    return ret_states


def add_uniform_transitions_to_hmm(model, states):
    state_count = len(states)
    dist = 1 / state_count

    for st_from in states:
        for st_to in states:
            model.add_transition(st_from, st_to, dist)
        model.add_transition(model.start, st_from, dist)


unique_syscalls = ['A', 'C', 'G', 'T']
unique_states = ['state1', 'state2', 'state3']
final_states = get_uniform_dist_for_all_states(unique_states, unique_syscalls)

model = HiddenMarkovModel('Attack classification')
model.add_states(final_states)
add_uniform_transitions_to_hmm(model, final_states)

model.bake()

seq1 = numpy.array(list('CGACTACTGACTACTCGCCGACGCGACTGCCGTCTATACTGCGCATACGGC'))
seq2 = numpy.array(list('CGACTACTCGCCGACGCATATATAGCGTGTATCACAGCCTGCGCATACGGC'))

X_train = [seq1, seq2]

model.fit(sequences=X_train, max_iterations=10, verbose=True)

print(model.dense_transition_matrix())
`

**Comment by mabuodeh** [comment]
I also add labels when fitting, but this is just an example, with data that I made up.

**Comment by mabuodeh** [comment]
I think I figured out the issue: when the transition matrix is initialized with a uniform distribution it doesn't change, and so I only changed the initial probability (model.start to the other states).
Should it do this though?

**Comment by jmschrei** [comment]
Sorry for the late reply. The issue is that when you set all the transitions AND the emissions to be uniform, the model cannot learn different distributions in an unsupervised manner. Imagine running k-means except that all of your initial centroids at the same point right in the middle. You'll see that your emissions are changing, but that they remain identical between states. Here is the output when you also print out the emissions distributions before and after training.

```
[{'A': 0.25, 'C': 0.25, 'G': 0.25, 'T': 0.25}]
[{'A': 0.25, 'C': 0.25, 'G': 0.25, 'T': 0.25}]
[1] Improvement: 2.8523510369665246	Time (s): 0.0002446
[2] Improvement: 5.684341886080802e-14	Time (s): 0.0002718
Total Training Improvement: 2.8523510369665814
Total Training Time (s): 0.0011
[{'A': 0.21568627450980407, 'C': 0.35294117647059126, 'G': 0.23529411764705788, 'T': 0.19607843137254688}]
[{'A': 0.21568627450980407, 'C': 0.35294117647059126, 'G': 0.23529411764705788, 'T': 0.19607843137254688}]
[[0.33333333 0.33333333 0.33333333 0.         0.        ]
 [0.33333333 0.33333333 0.33333333 0.         0.        ]
 [0.33333333 0.33333333 0.33333333 0.         0.        ]
 [0.33333333 0.33333333 0.33333333 0.         0.        ]
 [0.         0.         0.         0.         0.        ]]
```

You need a source of randomness either in the emissions or in the transitions. I'd recommend simply changing your `generate_uniform_distribution` function to be the following:

```python
def generate_uniform_distribution(observations):
    template_dist = {}
    
    random_probs = numpy.abs(numpy.random.randn(len(observations)))
    random_probs /= random_probs.sum()
    
    for obs, p in zip(observations, random_probs):
        template_dist[obs] = p

    return template_dist
```

When you do that, the outcome becomes this:

```
[{'A': 0.32765685909577097, 'C': 0.07432560283023468, 'G': 0.18179177873199662, 'T': 0.41622575934199774}]
[{'A': 0.47324500092405786, 'C': 0.24764525310833035, 'G': 0.24075535881685192, 'T': 0.03835438715075988}]
[1] Improvement: 16.284172173586626	Time (s): 0.000284
[2] Improvement: 0.9035406540911879	Time (s): 0.0002697
[3] Improvement: 0.9984491711572616	Time (s): 0.0002635
[4] Improvement: 1.1804914568137406	Time (s): 0.0002925
[5] Improvement: 1.3713260852897804	Time (s): 0.0002599
[6] Improvement: 1.4912192631509527	Time (s): 0.0002568
[7] Improvement: 1.4888950977405102	Time (s): 0.0002737
[8] Improvement: 1.4410158915371767	Time (s): 0.0002551
[9] Improvement: 1.5064189159796086	Time (s): 0.0002568
[10] Improvement: 1.6783069462667157	Time (s): 0.0002508
Total Training Improvement: 28.34383565561356
Total Training Time (s): 0.0034
[{'A': 0.1942371303789165, 'C': 0.02329133324599323, 'G': 0.5414054406853726, 'T': 0.2410660956897176}]
[{'A': 0.5020807929531226, 'C': 0.22275983552885903, 'G': 0.2722278530763032, 'T': 0.002931518441715248}]
[[0.0911075  0.53176114 0.37713136 0.         0.        ]
 [0.04757736 0.13115746 0.82126518 0.         0.        ]
 [0.59017099 0.35994179 0.04988722 0.         0.        ]
 [0.         0.00003288 0.99996712 0.         0.        ]
 [0.         0.         0.         0.         0.        ]]
```

**Comment by mabuodeh** [comment]
Yeah, I figured that was the issue. I'm following a count-based approach now, where the initial distribution is created from the data before baum welch. Thanks for explaining what the issue was!
